File: newBoCodeCodingChallenge.py
def odometer(arr, change):
  pos = len(arr) - 1
  while pos >= 0 and change != 0:
    sum = arr[pos] + change
    if sum < 0:
      sum = sum - 10
    change = int(sum / 10)
    arr[pos] = sum % 10
    pos = pos - 1
  if change == 1:
    arr[0] = 1

tests = [{"input": [4,3,9,5], "output": [4,3,9,6]}, {"input": [4,3,4,9], "output": [4,3,5,0]}, {"input": [9,9,9,9], "output": [1,0,0,0]}]

for test in tests:
  testArr = []
  for i in range(len(test["input"])):
    testArr.append(test["input"][i])
  odometer(testArr, 1)
  valid = True
  for i in range(len(testArr)):
    if testArr[i] != test["output"][i]:
      valid = False
  if valid:
    print(f"Test succeeded! {test['input']} => {test['output']}")
  else:
    print(f"Test failed! Expected {test['output']}, got {testArr}")

File: newBoCodeCodingChallenge.cs
using System;

class MainClass {
  public static void Main (string[] args) {
    int[] arr1 = {4, 3, 9, 5};
    Console.Write("Input: ");
    Console.WriteLine(string.Join(",", arr1));
    Odometer(arr1,1);
    Console.Write("Output: ");
    Console.WriteLine(string.Join(",", arr1));
    Console.WriteLine("Expected: 4,3,9,6");
    Console.WriteLine("----");

    int[] arr2 = {4, 3, 4, 9};
    Console.Write("Input: ");
    Console.WriteLine(string.Join(",", arr2));
    Odometer(arr2,1);
    Console.Write("Output: ");
    Console.WriteLine(string.Join(",", arr2));
    Console.WriteLine("Expected: 4,3,5,0");
    Console.WriteLine("----");

    int[] arr3 = {9, 9, 9, 9};
    Console.Write("Input: ");
    Console.WriteLine(string.Join(",", arr3));
    Odometer(arr3,1);
    Console.Write("Output: ");
    Console.WriteLine(string.Join(",", arr3));
    Console.WriteLine("Expected: 1,0,0,0");
    Console.WriteLine("----");
  }

  public static void Odometer (int[] arr, int change) {
    int pos = arr.Length - 1;
    while (pos >= 0 && change != 0) {
      int sum = arr[pos] + change;
      change = (int) Math.Floor(sum / 10d);
      arr[pos] = BetterMod(sum, 10);
      pos--;
    }
    if (change == 1) {
      arr[0] = 1;
    }
  }

  public static int BetterMod(int a, int b) {
    int modded = a % b;
    if (b > 0 && modded < 0) {
      while (modded < 0) {
        modded += b;
      }
    }
    return modded;
  }
}